src/compiler/checker.ts

```diff
@@ -6753,9 +6763,132 @@ namespace ts {
                     }
                 }
             }
+            if (!length(result) && indexWithLengthOverOne !== -1) {
+                // No sufficiently similar signature existed to subsume all the other signatures in the union - time to see if we can make a single
+                // signature that handles all over them. We only do this when there are overloads in only one constituent.
+                // (Overloads are conditional in nature and having overloads in multiple constituents would necessitate making a power set of
+                // signatures from the type, whose ordering would be non-obvious)
+                const masterList = signatureLists[indexWithLengthOverOne !== undefined ? indexWithLengthOverOne : 0];
+                let results: Signature[] | undefined = masterList.slice();
+                for (const signatures of signatureLists) {
+                    if (signatures === masterList) continue;
+                    const signature = signatures[0];
+                    Debug.assert(!!signature, "getUnionSignatures bails early on empty signature lists and should not have empty lists on second pass");
+                    results = mergeAsUnionSignature(results, signature);
+                    if (!results) {
+                        break;
+                    }
+                }
+                result = results;
+            }
             return result || emptyArray;
         }
 
+        function mergeAsUnionSignature(masterList: Signature[], addition: Signature): Signature[] | undefined {
+            if (addition.typeParameters && some(masterList, s => !!s.typeParameters)) {
+                return; // Can't currently unify type parameters
+            }
+            return map(masterList, sig => combineSignaturesOfUnionMembers(sig, addition));
+        }
+
+        function combineUnionThisParam(left: Symbol | undefined, right: Symbol | undefined): Symbol | undefined {
+            if (!left && !right) {
+                return;
+            }
+            if (!left) {
+                return right;
+            }
+            if (!right) {
+                return left;
+            }
+            // A signature `this` type might be a read or a write position... It's very possible that it should be invariant
+            // and we should refuse to merge signatures if there are `this` types and they do not match. However, so as to be
+            // permissive when calling, for now, we'll union the `this` types just like the overlapping-union-signature check does

```

Comment:
Let's talk about this. Not sure I understand why we'd treat `this` different than other parameters.

Reply:
Because we do in otherwise uniform union signatures.